src/server/main_service.cc

```diff
@@ -1042,8 +1068,24 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
   }
   DCHECK(cntx->transaction);
 
-  if (!eval_args.keys.empty())
+  bool scheduled = false;
+  int multi_mode = absl::GetFlag(FLAGS_multi_eval_mode);
+  DCHECK(multi_mode >= Transaction::GLOBAL && multi_mode <= Transaction::NON_ATOMIC);
+
+  if (multi_mode == Transaction::GLOBAL) {
+    scheduled = true;
+    cntx->transaction->StartMultiGlobal(cntx->db_index());
+  } else if (multi_mode == Transaction::LOCK_INCREMENTAL && !eval_args.keys.empty()) {
+    scheduled = true;
+    vector<bool> shards = DetermineKeyShards(eval_args.keys);
+    cntx->transaction->StartMultiLockedIncr(cntx->db_index(), shards);
+  } else if (multi_mode == Transaction::LOCK_AHEAD && !eval_args.keys.empty()) {
+    scheduled = true;
     cntx->transaction->StartMultiLockedAhead(cntx->db_index(), eval_args.keys);
+  } else if (multi_mode == Transaction::NON_ATOMIC) {
+    scheduled = true;
+    cntx->transaction->StartMultiNonAtomic();
+  };
```

Comment:
I don't like this part either.

A non-atomic multi transaction is neither a "classic" multi transaction neither a regular transaction because its re-initialized with the multi mechanism.

I had to decide whether to make it transparently multi and behave like one or make it explicitly non multi. I went with the first approach. It behaves exactly like a multi for the outside users.

Internally, there is `IsMulti` and `IsRunningMulti` where this distinguishment happens. What do you think?

Reply:
Maybe it should be renamed to smth like `IsAtomicMulti`?